Add ts Package, Time Series Key Encoding

mrtracy opened the pull request:
This commit creates the /ts package for handling time series data.

Included in the initial commit is the encoding and decoding logic for time
series data keys. This logic constructs cockroach keys based on the metric name
and timestamp for datapoints; this should providing a usefully ordered
distribution of time series data.

Commit:
Add ts Package, Time Series Key Encoding

This commit creates the /ts package for handling time series data.

Included in the initial commit is the encoding and decoding logic for time
series data keys. This logic constructs cockroach keys based on the series name
and timestamp for datapoints; this should providing a usefully ordered
distribution of time series data.

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2014 The Cockroach Authors.
```
`s/2014/2015/g`

mrtracy replied:
Done

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,27 @@
+/*
+Package ts provides a high-level time series API on top of an underlying
+cockroach datastore. Cockroach itself provides a single monolithic, sorted
+key-value map. Package ts collects time series data, consisting of datapoints
```
`s/datapoints/data points/g`

mrtracy replied:
Done

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,27 @@
+key-value map. Package ts collects time series data, consisting of datapoints
+with timestamps, and translates it into the key-value store.
+
+The ts package is a work in progress, and will initially only service queries
```
I'd add a `TODO(mtracy)` to this sentence to make it easily grep'able, otherwise the comment is likely to live past its expiration date.

mrtracy replied:
Done

petermattis commented on ts/keys.go:
```diff
@@ -0,0 +1,114 @@
+// resolutions are used to provide more detail for smaller timespans. The
+// resolution is denoted in the key by a three character string.
+//
+// The timestamp is the first timestamp for which this key might contain data.
```
I found this wording a bit confusing: "The timestamp is the first timestamp" and "this key". I imagine that you're rounding the timestamp to the resolution. 

This long comment is helpful, but you could make it better by providing an example or two of encoding a metric name and timestamp at 2 different resolutions.

petermattis commented on ts/keys_test.go:
```diff
@@ -0,0 +1,78 @@
+			t.Errorf("case %d, decoded timestamp %d did not match expected %d", i, a, e)
+		}
+		if a, e := r, tc.resolution; a != e {
+			t.Errorf("case %d, decoded resolution %d did not match expected %d", i, a, e)
```
I bet you could simplify the above by doing something like:

```
  d := tc
  d.name, d.source, d.resolution, d.timestamp = DecodeDataKey(encoded)
  if !reflect.DeepEqual(d, tc) {
    t.Errorf("case %d, decoded %+v != expected %+v", i, d, tc)
  }
```

mrtracy replied:
Done

petermattis commented on ts/keys.go:
```diff
@@ -0,0 +1,114 @@
+	KeyResolutionLength = 3
+	// KeyResolutionTenSeconds is the resolution key component for keys which
+	// store data with a ten second sample duration.
+	KeyResolutionTenSeconds = proto.Key("10s")
```
Do these vars need to be exported?

mrtracy replied:
These are going to be put directly into keys, which are visible in the key value store.  I think its a good candidate for exportation based on that alone.

petermattis replied:
Don't export it until you need to use it outside of this package. We currently export way to much stuff unnecessarily.

petermattis replied:
I think you should move the definition of `KeyResolutionTenSeconds` into `ts/resolution.go`.

petermattis commented on ts/keys.go:
```diff
@@ -0,0 +1,114 @@
+	timestamp int64) proto.Key {
+	return proto.MakeKey(KeyDataPrefix, encoding.EncodeBytes(nil, []byte(name)),
+		r.KeyComponent(), encoding.EncodeVarint(nil, timestamp),
+		encoding.EncodeBytes(nil, []byte(source)))
```
You don't need to use `encoding.EncodeBytes` for the last field in a key.

mrtracy replied:
Good point.

petermattis commented on ts/keys.go:
```diff
@@ -0,0 +1,114 @@
+		panic(fmt.Sprintf("malformed time series data key %v: no resolution", key))
+	}
+	resolution = ResolutionForKeyComponent(remainder[:KeyResolutionLength])
+	remainder = remainder[KeyResolutionLength:]
```
Curious why you are encoding the resolution as a string. I would have thought to encode it as a varint.

mrtracy replied:
I was just using the same strategy spencer used for local keys which are garbage collected.

petermattis commented on ts/resolution.go:
```diff
@@ -0,0 +1,102 @@
+
+const (
+	// Resolution10s stores data with a sample resolution of 10 seconds.
+	Resolution10s Resolution = iota
```
What other resolutions do you expect to define? This is all sorts of boring with only a single resolution.

mrtracy replied:
This will be used for rollups - blueflood, for example, does 5m, 20m, 1hr, 24hr rollups.
If we start storing external time series, we could potentially offer lower resolutions.

This information needs to go into the key before the timestamp, so I just needed to get it implemented now.

petermattis replied:
Now that you are encoding the `Resolution` value into the key, I think you should avoid using `iota` and manually specify the enum values. Also add a comment that the values must not change since they are stored in the encoded keys.

mrtracy replied:
Fixed and Added comment.

mrtracy commented:
Addressed all of the nits and suggestions, did a large overhaul of the documentation - put a lot of information in the document header outlining cockroach's strategy for storing time series data efficiently.  PTAL

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+these challenges.
+
+Cockroach organizes time series data by series name and timestamp; when querying
+a given series over a time range, all data in that range is thus sorted
```
`s/sorted/stored/g`

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+contiguously. However, cockroach transforms the data in a few important ways in
+order to effectively query a large amount of data.
+
+###Downsampling
```
Should there be a space between "###" and "Downsampling"? Not sure how this will look in godoc.

mrtracy replied:
You know, I was just assuming that godoc used markdown, but that probably was not a good assumption.  I want this to be a header; i'll look up exactly how to do it.

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+time a sample was taken is unimportant, with the overall trend of the data over
+time being far more important to analysis than the individual samples.
+
+With this in mind, Cockroach downsamples data before storing it; the original
```
You're inconsistent in your capitalization of "Cockroach". Probably best to always capitalize.

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+strategy helps reduce the number of keys scanned during a query.
+
+Finally, cockroach can record the same series at multiple sample durations, in a
+process commonly known as a "rollup".  For example, a single series may be
```
`s/a "rollup"/"rollup"/g`?

mrtracy replied:
Done.

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+together more efficiently.
+
+This is done by creating a "source key", an optional identifier that is separate
+from the series name itself. Source keys are sorted *after* series name and
```
"sorted *after*" reads awkwardly to me. I think it would be clearer to say that source keys are used as the suffix of the key or are appended to the key after the series name and timestamp.

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+	tsd.cockroach.disksize.ExampleApplication.10s.1451646000.9
+	tsd.cockroach.disksize.ExampleApplication.10s.1451646000.10
+
+Data for all each source is stored in two keys: one for the 10 pm hour, and one
```
`s/all each source/each source/g`

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+keys will appear sorted in the data store as shown.
+
+(Note that the keys will NOT be exactly as pictured above; they will be encoded
+in a way that is more efficient, but may not be readily human readable.)
```
`s/but may not be/but is not/g`

petermattis commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+// Author: Matt Tracy (matt.r.tracy@gmail.com)
+
+/*
+Package ts provides a high-level time series API on top of an underlying
```
This is great. Love having the overall design embedded in the package documentation comment.

petermattis commented on ts/keys.go:
```diff
@@ -0,0 +1,95 @@
+	timestamp int64) proto.Key {
+	return proto.MakeKey(keyDataPrefix, encoding.EncodeBytes(nil, []byte(name)),
+		encoding.EncodeVarint(nil, int64(r)), encoding.EncodeVarint(nil, timestamp),
+		[]byte(source))
```
You really should be building up the key incrementally. That is why `encoding.EncodeBytes` takes a slice to append to. Something like:

```
  k := append([]byte(nil), keyDataPrefix...)
  k = encoding.EncodeBytes(k, []byte(name))
  k = encoding.EncodeVarint(k, int64(r))
  k = encoding.EncodeVarint(k, timestamp)
  k = append(k, source...)
  return proto.Key(k)
```

Does a particular resolution imply a particular key duration? If yes, should `MakeDataKey` be normalizing the timestamp? For example, `timestamp /= r.KeyDuration()`.

mrtracy replied:
I thought that passing them all to MakeKey might use less memory.

Good point about the KeyDuration; I was expecting the timestamp to already be truncated at this point, but there's no reason not to normalize it here as well.

petermattis replied:
Using `proto.MakeKey` will actually use more memory. Heck, casting the `name` (a `string`) to a `[]byte` will allocate memory. Not worth worrying about (yet), but it is good to now what is going on.

petermattis commented on ts/resolution.go:
```diff
@@ -0,0 +1,68 @@
+
+// SampleDuration returns the sample duration corresponding to this resolution
+// value, expressed in nanoseconds.
+func (r Resolution) SampleDuration() int64 {
```
Not sure if `SampleDuration` or `KeyDuration` should be exported, but ok to leave as-is for now.

petermattis commented:
LGTM

tschottdorf commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+
+Cockroach organizes time series data by series name and timestamp; when querying
+a given series over a time range, all data in that range is thus sorted
+contiguously. However, cockroach transforms the data in a few important ways in
```
are we capitalizing Cockroach or not?

tschottdorf commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+
+A specific sample duration in cockroach is known as a Resolution. Cockroach
+supports a fixed set of Resolutions; each Resolution has a fixed sample duration
+and a key duration. For example, the resolution "Resolution10s" has a sample
```
Should the name "Resolution10s" also allude to the one hour key duration?

mrtracy replied:
It might, but Resolution10s1h looked kind of odd to me... I'll make this decision when we introduce another resolution.

tschottdorf commented on ts/doc.go:
```diff
@@ -0,0 +1,139 @@
+The series name is: cockroach.disksize.ExampleApplication
+
+Data points for this series are automatically collected from any store that
+contains ranges with a prefix of "ExampleApplication". When data points are
```
s/ranges/keys/

mrtracy commented:
I believe I addressed all comment nits.  Changed MakeKey to encode timestamp keys incrementally, and I now normalize the timestamp inside of this method.  It is now actually normalized to a "time slot", which reduces the key sizes a bit from full-length nanosecond timestamps.  I've updated the comments to reflect this.

petermattis commented:
Looks good.

spencerkimball commented on ts/keys.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+// Time series keys are carefully constructed to usefully sort the data in the
+// key-value store for the purpose of queries. Time series data is queryied by
```
s/queryied/queried/

spencerkimball commented on ts/keys.go:
```diff
@@ -0,0 +1,108 @@
+// timestamp to an exact multiple of the key duration, and then dividing it by
+// the key duration:
+//
+// 		slot := (timestamp / keyDuration) // integer division
```
Is timestamp measured in nanoseconds or seconds? Specify that here.

spencerkimball commented on ts/resolution.go:
```diff
@@ -0,0 +1,70 @@
+// sampleDurationByResolution is a map used to retrieve the sample duration
+// corresponding to a Resolution value. Sample durations are expressed in
+// nanoseconds.
+var sampleDurationByResolution = map[Resolution]int64{
```
Seems kind of annoying to have a map for this. Why not just make the value of the enum the number of nanoseconds in the duration?

    const (
      Resolution10s Resolution = (10 * time.Second).Nanoseconds()
    )

    func (r Resolution) SampleDuration() int64 { return int64(r) }

petermattis replied:
The value of `Resolution` is encoded into the key, so this kind of sucks. If we're worried about performance here (which we shouldn't be yet), all of these maps could be slices:

```
var sampleDurationByResolution = [...]int64{
  Resolution10s: int64(time.Second * 10)
}
```

spencerkimball commented:
LGTM
